core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -927,64 +957,55 @@ private void lockDescendantsInternal(LockedInodePath inodePath, LockMode lockMod
    */
   public void deleteInode(RpcContext rpcContext, LockedInodePath inodePath, long opTimeMs,
       DeleteOptions deleteOptions) throws FileDoesNotExistException {
-    Inode<?> inode = inodePath.getInode();
-    InodeDirectory parent = (InodeDirectory) mInodes.getFirst(inode.getParentId());
+    InodeView inode = inodePath.getInode();
+    InodeDirectoryView parent = (InodeDirectoryView) mInodes.getById(inode.getParentId());
     if (parent == null) {
       LOG.warn("Parent id not found: {} deleting inode: {}", inode.getParentId(), inode);
       throw new FileDoesNotExistException(
           ExceptionMessage.INODE_DOES_NOT_EXIST.getMessage(inode.getParentId()));
     }
 
-    // Journal before removing the inode from the parent, since the parent is read locked.
```

Comment:
yes, and we have a wall of text to explain why :D

```
    // Unlike most entries, the delete file entry must be applied *before* making the in-memory
    // change. This is because delete file and create file are performed with only a read lock on
    // the parent directory. As soon as we do the in-memory-delete, another thread could re-create a
    // directory with the same name, and append a journal entry for the inode creation. This would
    // ruin journal replay because we would see two create file entries in a row for the same file
    // name. The opposite order is safe. We will never append the delete entry for a file before its
    // creation entry because delete requires a write lock on the deleted file, but the create
    // operation holds that lock until after it has appended to the journal.
    context.get().append(JournalEntry.newBuilder().setDeleteFile(entry).build());
```